pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_bastion.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	toolscache "k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+)
+
+func (g *graph) setupBastionWatch(informer cache.Informer) {
+	informer.AddEventHandler(toolscache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			bastion, ok := obj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+			g.handleBastionCreateOrUpdate(bastion)
+		},
+
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldBastion, ok := oldObj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+
+			newBastion, ok := newObj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+
+			if !apiequality.Semantic.DeepEqual(oldBastion.Spec.SeedName, newBastion.Spec.SeedName) {
+				g.handleBastionCreateOrUpdate(newBastion)
+			}
+		},
+
+		DeleteFunc: func(obj interface{}) {
+			if tombstone, ok := obj.(toolscache.DeletedFinalStateUnknown); ok {
+				obj = tombstone.Obj
+			}
+			bastion, ok := obj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+			g.handleBastionDelete(bastion)
+		},
+	})
+}
+
+func (g *graph) handleBastionCreateOrUpdate(bastion *gardencorev1alpha1.Bastion) {
+	start := time.Now()
+	defer func() {
+		metricUpdateDuration.WithLabelValues("Bastion", "CreateOrUpdate").Observe(time.Since(start).Seconds())
+	}()
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	g.deleteVertex(VertexTypeBastion, bastion.Namespace, bastion.Name)
+
+	bastionVertex := g.getOrCreateVertex(VertexTypeBastion, bastion.Namespace, bastion.Name)
```

Comment:
Nit: This could be moved into the `if {}` block

Reply:
Technically, now that we forbid creating Bastions for Shoots that have no Seed yet, we could remove the entire `if` statement? WDYT?

Reply:
Hmm, not sure, we have the control plane migration use-case where a `Shoot`'s `.spec.seedName` can change. It's not clear to me what would happen to `Bastion`s in this case. Probably it should be simply terminated immediately instead of "moving" it as well. So yes, we could remove this statement most likely. OTOH, it doesn't harm or produce maintenance effort. The benefit of keeping it would be to not include implementation details of the API validation, for example, so it's easier to "not miss" it in case above description changes. Anyways, please do as you like - I don't have a strong opinion on this.